pkg/provisioner/templates/containerd.go

```diff
@@ -32,29 +32,121 @@ const containerdTemplate = `
 with_retry 3 10s sudo apt-get update
 install_packages_with_retry ca-certificates curl gnupg -y
 sudo install -m 0755 -d /etc/apt/keyrings
-curl -fsSL https://download.docker.com/linux/ubuntu/gpg | sudo gpg --dearmor -o /etc/apt/keyrings/docker.gpg
-sudo chmod a+r /etc/apt/keyrings/docker.gpg
 
-# Add the repository to Apt sources:
-echo \
-  "deb [arch="$(dpkg --print-architecture)" signed-by=/etc/apt/keyrings/docker.gpg] https://download.docker.com/linux/ubuntu \
-  "$(. /etc/os-release && echo "$VERSION_CODENAME")" stable" | \
-  sudo tee /etc/apt/sources.list.d/docker.list > /dev/null
-  with_retry 3 10s sudo apt-get update
+# Check if CONTAINERD_VERSION is empty, if so fetch the latest stable version
+if [ -z "$CONTAINERD_VERSION" ]; then
+    echo "Fetching latest stable containerd version..."
+    CONTAINERD_VERSION=$(curl -fsSL https://api.github.com/repos/containerd/containerd/releases/latest | grep '"tag_name":' | cut -d '"' -f 4 | sed 's/v//')
 
-# Install containerd
-install_packages_with_retry containerd.io=${CONTAINERD_VERSION}-1
+    if [ -z "$CONTAINERD_VERSION" ]; then
+        echo "Failed to fetch latest Containerd version. Exiting."
+        exit 1
+    fi
+fi
+echo "Using containerd version: $CONTAINERD_VERSION"
 
-# Configure containerd and start service
-mkdir -p /etc/containerd
+ARCH=$(uname -m)
+CONTAINERD_TAR="containerd-${CONTAINERD_VERSION}-linux-${ARCH}.tar.gz"
+CONTAINERD_URL="https://github.com/containerd/containerd/releases/download/v${CONTAINERD_VERSION}/${CONTAINERD_TAR}"
+CONTAINERD_SHA256_URL="https://github.com/containerd/containerd/releases/download/v${CONTAINERD_VERSION}/SHA256SUMS"
+
+echo "Downloading and extracting containerd ${CONTAINERD_VERSION} from: $CONTAINERD_URL"
+
+# Download containerd tarball and checksum
+curl -fsSL -o ${CONTAINERD_TAR} ${CONTAINERD_URL}
+curl -fsSL -o containerd_SHA256SUMS ${CONTAINERD_SHA256_URL}
+
+# Verify SHA256 checksum
+EXPECTED_CHECKSUM=$(grep ${CONTAINERD_TAR} containerd_SHA256SUMS | awk '{print $1}')
+ACTUAL_CHECKSUM=$(sha256sum ${CONTAINERD_TAR} | awk '{print $1}')
+
+if [ "$EXPECTED_CHECKSUM" != "$ACTUAL_CHECKSUM" ]; then
+    echo "Checksum verification failed for containerd. Exiting."
+    exit 1
+fi
+
+# Stream directly into tar to avoid saving the archive
+cat ${CONTAINERD_TAR} | sudo tar Cxzvf /usr/local -
```

Comment:
We're already downloading the archive. This is only needed if we're using `curl` directly. Since we're verifying the SHA that's fine too.

Not a blocker though.